### ISSUE TITLE
EVSS VSOSearch returns authentication failed

### ISSUE THREAD

**Issue by shawkey-daniel** [department-of-veterans-affairs/va.gov-team#22732]
## Issue Description
The vsa-education team has a requirement to retrieve the userPoaInfoAvailable attribute from the EVSS VSOSearch service in support of the STEM scholarship automated denial. The request is made [here](https://github.com/department-of-veterans-affairs/vets-api/blob/99a1a197d3207c8a81c951b1568ae6e1cc54642c/app/workers/education_form/process10203_submissions.rb#L89). The errors are logged [here](http://sentry.vfs.va.gov/organizations/vsp/issues/31328/?environment=production&project=3&query=is%3Aunresolved+%22Failed+to+retrieve+VSOSearch+data%22&statsPeriod=14d). The errors indicate EDIPI is required for the request. EDIPI should be available.

---
## Tasks
- [ ] Figure out why EDIPI is not available or how to retrieve it or change the service to not require EDIPI.

## Acceptance Criteria
- [ ] Resolve issue retrieving userHasPoaAvailable
---
## How to configure this issue
- [ ] **Attached to a Milestone** (when will this be completed?)
- [ ] **Attached to an Epic** (what body of work is this a part of?)
- [ ] **Labeled with Team** (`product support`, `analytics-insights`, `operations`, `service-design`, `tools-be`, `tools-fe`)
- [ ] **Labeled with Practice Area** (`backend`, `frontend`, `devops`, `design`, `research`, `product`, `ia`, `qa`, `analytics`, `contact center`, `research`, `accessibility`, `content`)
- [ ] **Labeled with Type** (`bug`, `request`, `discovery`, `documentation`, etc.)

**Comment by bramleyjl** [comment]
After investigating some we think these issues might be happening with users that are missing LOA3 authentication and/or an MHV ICN. The method to find a user's EDIPI is:

``` 
def edipi
    loa3? && dslogon_edipi.present? ? dslogon_edipi : mpi&.edipi
end
```

So the method will check to see if the user has LOA3 & an EDIPI from DSLogon and use that if they both exist. If not, it will attempt to get the EDIPI from MPI. However, the user still needs to have LOA3 in order to have access to the MPI profile. In addition, in order to look up the user in MPI we need to already have either the EDIPI from DSLogon or the MHV ICN, and we're unlikely to have the former in this circumstance.

To sum up: If the user is missing LOA3 we cannot look up their EDIPI. If they are missing an MHV ICN as well as a DSLogon EDIPI then we will also be unable to look up their EDIPI.

The PR above adds a Sentry log line for when EDIPI is missing within `request_headers`, so once it goes line we will have more information on whether one or both of these missing attributes is responsible.

**Comment by bramleyjl** [comment]
Additional investigation and logging has lead us to believe that the issues are stemming from [this PR](https://github.com/department-of-veterans-affairs/vets-api/pull/5793/files) that changed the EVSS service from being instantiated with a `User` instance to using authentication_headers. In [the new logic](https://github.com/department-of-veterans-affairs/vets-api/blob/7658f1eb158cdb9e5203d814978862172ea04cbb/lib/evss/service.rb#L12), the `user` is still passed in to serve as a backup source of information for the service, but in the place where [the service is invoked to find the user's PoA information](https://github.com/department-of-veterans-affairs/vets-api/blob/7658f1eb158cdb9e5203d814978862172ea04cbb/app/workers/education_form/process10203_submissions.rb#L91) no user is passed in which means the service only has the `auth_headers` to derive the necessary user attributes from.

The original error was due to the `auth_headers` possessing a "valid" key-value pair for `va_eauth_dodedipnid` where the value was actually an empty string, thus the backup check of the `user`'s edipi wasn't occurring. My most recent change to [add additional logging](https://github.com/department-of-veterans-affairs/vets-api/pull/7009/files#diff-0e63929d51b54a77392e6a64d1442edec3f66cb3903ecd1e07902b61fedeb637) also fixed that non-functioning conditional, so when the new logic found that the `auth_headers` `va_eauth_dodedipnid` wasn't actually present and attempted to get the value from the `user` an [undefined method error from calling `ssn`/`edipi` on the missing `user` resulted](http://sentry.vfs.va.gov/organizations/vsp/issues/38635/events/oldest/?project=3&query=is%3Aunresolved&statsPeriod=14d).

As such, there are two separate fixes that need to occur here:
* The `additional_headers` need to be fixed such that they are not including an empty string as the value for an EDIPI or SSN, and all checks to determine whether to use additional_headers/user should be updated to make sure they evaluate an empty string as `false`.
* The EVSS service instantiation needs to be fixed to properly take a user as an argument so that the user's attribute values will be available as backup values in the event that `additional_headers' is missing an attribute value.

**Comment by bramleyjl** [comment]
PR #22732 is intended as a temporary fix, falling back to a passed-through `Account` object to obtain a user's EDIPI if it is not present in the auth_headers. The same could not be done for a user's SSN, which leads to the real issue with the current implementation of this Sidekiq job; it relies on information that is ultimately derived from the User object, which is only temporarily stored in Redis and will most likely not be available when the job actually runs. On top of this, there currently is no good way to call for protected user information like an SSN without the user being LOA3, an impossibility if this is all being run as part of a Sidekiq job.

After further discussion with @bosawt we agreed that a potential solution for this issue would be for the Sidekiq job to be updated to use user EDIPIs and SSNs that encrypted and stored at the time of the job's creation when User object needed to obtain those values should be available. It appears that there was the beginning of an effort towards something along these lines at the beginning of 2021, when the `education_stem_automated_decisions` table was added to the vets-api schema with an `encrypted_auth_headers_json` column, presumably for this purpose?

**Comment by shawkey-daniel** [comment]
That is why we were saving the auth headers, so we could try to access the EVSS service with the auth headers. Unfortunately, it never worked. The errors indicated the EDIPI was still missing. That is why this ticket was created.

**Comment by bramleyjl** [comment]
After reviewing Sentry it appears that the missing EDIPI errors are continuing; I'll take another look at what is causing this in Sprint 55.

**Comment by bramleyjl** [comment]
@dmccurdybah and did some pairing on this on 7/2 but were blocked by Sentry being down. As of now Sentry is still not functioning for me, so I'm marking this blocked until it becomes functional again.

**Comment by bramleyjl** [comment]
After extensive effort @dmccurdybah  and I were unable to determine the source of the EVSS service's 502 error. Given the presence of higher-priority items for both the Education and Identity teams, the rotation of the Education team off of VSP in the near future, and the fact that the affected feature is currently disabled behind a feature flag, we have decided to halt further investigations of this issue. Once the new team gets integrated and up to speed, if they decide to re-open this ticket I will be happy to share with them what Devin and I have learned so far and to provide ongoing support.
